arangod/Aql/AllRowsFetcher.h

```diff
@@ -0,0 +1,97 @@
+////////////////////////////////////////////////////////////////////////////////
+/// DISCLAIMER
+///
+/// Copyright 2018 ArangoDB GmbH, Cologne, Germany
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Tobias Gödderz
+////////////////////////////////////////////////////////////////////////////////
+
+#ifndef ARANGOD_AQL_ALL_ROWS_FETCHER_H
+#define ARANGOD_AQL_ALL_ROWS_FETCHER_H
+
+#include "Aql/AqlItemMatrix.h"
+#include "Aql/ExecutionState.h"
+
+#include <memory>
+
+namespace arangodb {
+namespace aql {
+
+class AqlItemBlock;
+class BlockFetcher;
+
+/**
+ * @brief Interface for all AqlExecutors that do need all
+ *        rows at a time in order to make progress.
+ *        The guarantee is the following:
+ *        If fetchAllRows returns a matrix the pointer to
```

Comment:
This describes an unique_ptr. If we could access the engine globally it could be put in an custom deleter.